File: shared/src/main/java/ru/privatenull/pnauth/command/AuthCommandRequest.java
package ru.privatenull.pnauth.command;

import java.util.List;
import java.util.UUID;
import java.util.function.Predicate;

public record AuthCommandRequest(
        UUID uniqueId,
        String username,
        String command,
        List<String> arguments,
        Predicate<String> permissionChecker
) {
    public AuthCommandRequest(UUID uniqueId, String username, String command, List<String> arguments) {
        this(uniqueId, username, command, arguments, permission -> false);
    }

    public AuthCommandRequest {
        arguments = arguments == null ? List.of() : List.copyOf(arguments);
        command = command == null ? "" : command;
        permissionChecker = permissionChecker == null ? permission -> false : permissionChecker;
    }

    public boolean isPlayer() {
        return uniqueId != null;
    }

    public boolean hasPermission(String permission) {
        return permissionChecker.test(permission);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/command/AuthCommandRequest.kt
package ru.privatenull.pnauth.command

import java.util.UUID
import java.util.function.Predicate

@JvmRecord
data class AuthCommandRequest(
    val uniqueId: UUID?,
    val username: String?,
    val command: String = "",
    val arguments: List<String> = emptyList(),
    val permissionChecker: Predicate<String> = Predicate { false }
) {
    constructor(
        uniqueId: UUID?,
        username: String?,
        command: String?,
        arguments: List<String>?
    ) : this(
        uniqueId,
        username,
        command ?: "",
        arguments?.toList() ?: emptyList(),
        Predicate { false }
    )

    fun isPlayer(): Boolean = uniqueId != null

    fun hasPermission(permission: String): Boolean = permissionChecker.test(permission)
}
